app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,212 @@
 package org.commcare.utils;
 
+import android.os.Build;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+
+import androidx.annotation.RequiresApi;
+
 import org.commcare.util.Base64;
+import org.commcare.util.Base64DecoderException;
 
+import java.io.IOException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableEntryException;
+import java.security.cert.CertificateException;
+import java.util.Random;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
 
 /**
  * Utility class for encrypting submissions during the SaveToDiskTask.
  *
  * @author mitchellsundt@gmail.com
  */
+
 public class EncryptionUtils {
 
+    private static final int PASSPHRASE_LENGTH = 32;
+    private static final String SECRET_NAME = "secret";
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String ALGORITHM = KeyProperties.KEY_ALGORITHM_AES;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+    private static KeyStore keystore = null;
+
+    //Gets the SecretKey from the Android KeyStore (creates a new one the first time)
+    private static SecretKey getKey()
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException,
+            UnrecoverableEntryException, InvalidAlgorithmParameterException {
+        if(keystore == null) {
+            keystore = KeyStore.getInstance("AndroidKeyStore");
+            keystore.load(null);
+        }
+
+        KeyStore.Entry existingKey = keystore.getEntry(SECRET_NAME, null);
```

Comment:
Does this key ever change ? For eg. if we clear the app data on the phone.

Reply:
My understanding [from](https://stackoverflow.com/questions/47343659/does-clear-data-on-app-settings-clears-all-android-keystore-stored-for-that-cert) here is that the keystore data for an app is erased when the user clears all app data. But since the DBs are also erased in that case, a new DB will have to be created and it's okay for a different key to be generated and used going forward.